File: nim/src/midi2chiptune.nim
import std/[os, streams, endians, tables, sequtils, math, strutils, algorithm, options]

# WAVファイルのヘッダー構造体
type
  WaveHeader = object
    chunkID: array[4, char]        # "RIFF"
    chunkSize: uint32
    format: array[4, char]         # "WAVE"
    subchunk1ID: array[4, char]    # "fmt "
    subchunk1Size: uint32
    audioFormat: uint16
    numChannels: uint16
    sampleRate: uint32
    byteRate: uint32
    blockAlign: uint16
    bitsPerSample: uint16
    subchunk2ID: array[4, char]    # "data"
    subchunk2Size: uint32

# MIDIイベント
type
  MidiEventKind = enum
    NoteOn, NoteOff, ProgramChange, Unknown
  
  MidiEvent = object
    case kind: MidiEventKind
    of NoteOn:
      noteOnChannel: int
      noteOnNote: int
      velocity: int
      noteOnTime: int
    of NoteOff:
      noteOffChannel: int
      noteOffNote: int
      noteOffTime: int
    of ProgramChange:
      progChannel: int
      program: int
      progTime: int
    of Unknown:
      discard

# MIDIトラック
type
  MidiTrack = object
    events: seq[MidiEvent]

# MIDIファイル
type
  MidiFile = object
    format: int
    trackCount: int
    ticksPerQuarter: int
    tracks: seq[MidiTrack]

# 音符情報
type
  Note = object
    midiNote: int
    channel: int
    startTime: float
    duration: float
    velocity: int

# 矩形波のデューティサイクル（ファミコンの4種類）
type
  DutyCycle = enum
    Duty12_5  # 12.5%
    Duty25    # 25%
    Duty50    # 50%
    Duty75    # 75%

# デューティサイクルの値を取得
proc getDutyCycleValue(duty: DutyCycle): float =
  case duty
  of Duty12_5: 0.125
  of Duty25: 0.25
  of Duty50: 0.5
  of Duty75: 0.75

# MIDIノート番号を周波数に変換
proc midiNoteToFrequency(midiNote: int): float =
  440.0 * pow(2.0, float(midiNote - 69) / 12.0)

# Variable Length Quantity (VLQ) の読み込み
proc readVLQ(stream: Stream): int =
  var value = 0
  var byte: uint8
  
  while true:
    byte = stream.readUint8()
    value = (value shl 7) or int(byte and 0x7F)
    if (byte and 0x80) == 0:
      break
  
  result = value

# ビッグエンディアンで16ビット整数を読み込み
proc readBigEndian16(stream: Stream): uint16 =
  var bytes: array[2, uint8]
  discard stream.readData(addr bytes, 2)
  result = (uint16(bytes[0]) shl 8) or uint16(bytes[1])

# ビッグエンディアンで32ビット整数を読み込み
proc readBigEndian32(stream: Stream): uint32 =
  var bytes: array[4, uint8]
  discard stream.readData(addr bytes, 4)
  result = (uint32(bytes[0]) shl 24) or (uint32(bytes[1]) shl 16) or 
           (uint32(bytes[2]) shl 8) or uint32(bytes[3])

# MIDIヘッダーチャンクの読み込み
proc readMidiHeader(stream: Stream): (int, int, int) =
  var chunkType: array[4, char]
  discard stream.readData(addr chunkType, 4)
  
  let chunkLength = stream.readBigEndian32()
  let format = stream.readBigEndian16()
  let trackCount = stream.readBigEndian16()
  let ticksPerQuarter = stream.readBigEndian16()
  
  (int(format), int(trackCount), int(ticksPerQuarter))

# MIDIイベントの解析
proc parseMidiEvent(stream: Stream, runningStatus: var int, deltaTime: int): MidiEvent =
  var status = runningStatus
  let firstByte = stream.readUint8()
  
  if firstByte >= 128:
    status = int(firstByte)
  else:
    stream.setPosition(stream.getPosition() - 1)
  
  let statusType = status and 0xF0
  
  case statusType
  of 0x90: # Note On
    let channel = status and 0x0F
    let note = int(stream.readUint8())
    let velocity = int(stream.readUint8())
    runningStatus = status
    if velocity == 0:
      return MidiEvent(kind: NoteOff, noteOffChannel: channel, noteOffNote: note, noteOffTime: deltaTime)
    else:
      return MidiEvent(kind: NoteOn, noteOnChannel: channel, noteOnNote: note, velocity: velocity, noteOnTime: deltaTime)
  
  of 0x80: # Note Off
    let channel = status and 0x0F
    let note = int(stream.readUint8())
    discard stream.readUint8() # velocity
    runningStatus = status
    return MidiEvent(kind: NoteOff, noteOffChannel: channel, noteOffNote: note, noteOffTime: deltaTime)
  
  of 0xC0: # Program Change
    let channel = status and 0x0F
    let program = int(stream.readUint8())
    runningStatus = status
    return MidiEvent(kind: ProgramChange, progChannel: channel, program: program, progTime: deltaTime)
  
  else:
    # その他のイベントはスキップ
    if status == 0xFF: # Meta event
      discard stream.readUint8() # event type
      let length = readVLQ(stream)
      for _ in 0..<length:
        discard stream.readUint8()
    elif status >= 0x80:
      # その他のMIDIイベント
      discard stream.readUint8()
      if (status and 0xE0) != 0xC0 and (status and 0xE0) != 0xD0:
        discard stream.readUint8()
    
    runningStatus = status
    return MidiEvent(kind: Unknown)

# MIDIトラックの読み込み
proc readMidiTrack(stream: Stream): MidiTrack =
  var chunkType: array[4, char]
  discard stream.readData(addr chunkType, 4)
  
  let chunkLength = stream.readBigEndian32()
  let endPosition = stream.getPosition() + int64(chunkLength)
  
  var events: seq[MidiEvent] = @[]
  var runningStatus = 0
  
  while stream.getPosition() < endPosition:
    try:
      let deltaTime = readVLQ(stream)
      let event = parseMidiEvent(stream, runningStatus, deltaTime)
      if event.kind != Unknown:
        events.add(event)
    except:
      if stream.getPosition() < endPosition:
        discard stream.readUint8()
  
  MidiTrack(events: events)

# MIDIファイル全体の読み込み
proc readMidiFile(filename: string): Option[MidiFile] =
  try:
    let stream = newFileStream(filename, fmRead)
    defer: stream.close()
    
    let (format, trackCount, ticksPerQuarter) = readMidiHeader(stream)
    
    var tracks: seq[MidiTrack] = @[]
    for i in 0..<trackCount:
      tracks.add(readMidiTrack(stream))
    
    some(MidiFile(
      format: format,
      trackCount: trackCount,
      ticksPerQuarter: ticksPerQuarter,
      tracks: tracks
    ))
  except:
    echo "MIDIファイルの読み込みエラー: ", getCurrentExceptionMsg()
    none(MidiFile)

# MIDIイベントから音符リストに変換
proc eventsToNotes(midiFile: MidiFile, tempo: float): seq[Note] =
  proc ticksToSeconds(tick: int): float =
    float(tick) / float(midiFile.ticksPerQuarter) * 60.0 / tempo
  
  var noteOnEvents = initTable[(int, int), (int, float)]() # (channel, note) -> (velocity, startTime)
  var notes: seq[Note] = @[]
  
  for track in midiFile.tracks:
    var currentTick = 0
    for event in track.events:
      case event.kind
      of NoteOn:
        currentTick += event.noteOnTime
        let startTime = ticksToSeconds(currentTick)
        noteOnEvents[(event.noteOnChannel, event.noteOnNote)] = (event.velocity, startTime)
      
      of NoteOff:
        currentTick += event.noteOffTime
        let endTime = ticksToSeconds(currentTick)
        let key = (event.noteOffChannel, event.noteOffNote)
        if key in noteOnEvents:
          let (velocity, startTime) = noteOnEvents[key]
          let duration = endTime - startTime
          if duration > 0.0:
            notes.add(Note(
              midiNote: event.noteOffNote,
              channel: event.noteOffChannel,
              startTime: startTime,
              duration: duration,
              velocity: velocity
            ))
          noteOnEvents.del(key)
      
      else:
        discard
  
  notes

# 波形生成関数
proc generateSquareWave(frequency: float, duty: DutyCycle, sampleRate: int, duration: float): seq[float] =
  let samples = int(float(sampleRate) * duration)
  let dutyValue = getDutyCycleValue(duty)
  
  result = newSeq[float](samples)
  for i in 0..<samples:
    let t = float(i) / float(sampleRate)
    let phase = (t * frequency) mod 1.0
    if phase < dutyValue:
      result[i] = 1.0
    else:
      result[i] = -1.0

proc generateTriangleWave(frequency: float, sampleRate: int, duration: float): seq[float] =
  let samples = int(float(sampleRate) * duration)
  
  result = newSeq[float](samples)
  for i in 0..<samples:
    let t = float(i) / float(sampleRate)
    let phase = (t * frequency) mod 1.0
    if phase < 0.5:
      result[i] = 4.0 * phase - 1.0
    else:
      result[i] = 3.0 - 4.0 * phase

proc generateNoise(isShort: bool, sampleRate: int, duration: float): seq[float] =
  let samples = int(float(sampleRate) * duration)
  var shift = 1'u16
  
  result = newSeq[float](samples)
  for i in 0..<samples:
    let bit0 = shift and 1
    let bit1 = if isShort: (shift shr 6) and 1 else: (shift shr 1) and 1
    let feedback = bit0 xor bit1
    shift = (shift shr 1) or (feedback shl 14)
    
    if bit0 == 1:
      result[i] = 1.0
    else:
      result[i] = -1.0

# 音符を波形に変換
proc noteToWaveform(note: Note, sampleRate: int, totalDuration: float): seq[float] =
  let frequency = midiNoteToFrequency(note.midiNote)
  let volume = float(note.velocity) / 127.0 * 0.7
  let startSample = int(note.startTime * float(sampleRate))
  let noteSamples = int(note.duration * float(sampleRate))
  let totalSamples = int(totalDuration * float(sampleRate))
  
  let waveform = case note.channel mod 4
    of 0: generateSquareWave(frequency, Duty50, sampleRate, note.duration)
    of 1: generateSquareWave(frequency, Duty25, sampleRate, note.duration)
    of 2: generateTriangleWave(frequency, sampleRate, note.duration)
    else: generateNoise(false, sampleRate, note.duration)
  
  result = newSeq[float](totalSamples)
  let endSample = min(startSample + noteSamples, totalSamples)
  
  for i in startSample..<endSample:
    if i - startSample < waveform.len:
      result[i] = waveform[i - startSample] * volume

# 複数の波形をミックス
proc mixWaveforms(waveforms: seq[seq[float]]): seq[float] =
  if waveforms.len == 0:
    return @[]
  
  let maxLength = waveforms.mapIt(it.len).max()
  result = newSeq[float](maxLength)
  
  for i in 0..<maxLength:
    var sum = 0.0
    for w in waveforms:
      if i < w.len:
        sum += w[i]
    result[i] = sum / float(waveforms.len)

# 16ビット整数に変換
proc convertTo16Bit(samples: seq[float]): seq[int16] =
  result = newSeq[int16](samples.len)
  for i, sample in samples:
    let clamped = max(-1.0, min(1.0, sample))
    result[i] = int16(clamped * 32767.0)

# WAVヘッダー作成
proc createWaveHeader(sampleRate, numChannels, bitsPerSample, dataSize: int): WaveHeader =
  let byteRate = uint32(sampleRate * numChannels * bitsPerSample div 8)
  let blockAlign = uint16(numChannels * bitsPerSample div 8)
  
  result.chunkID = ['R', 'I', 'F', 'F']
  result.chunkSize = uint32(36 + dataSize)
  result.format = ['W', 'A', 'V', 'E']
  result.subchunk1ID = ['f', 'm', 't', ' ']
  result.subchunk1Size = 16
  result.audioFormat = 1
  result.numChannels = uint16(numChannels)
  result.sampleRate = uint32(sampleRate)
  result.byteRate = byteRate
  result.blockAlign = blockAlign
  result.bitsPerSample = uint16(bitsPerSample)
  result.subchunk2ID = ['d', 'a', 't', 'a']
  result.subchunk2Size = uint32(dataSize)

# WAVファイル書き込み
proc writeWaveFile(filename: string, samples: seq[float], sampleRate: int) =
  let stream = newFileStream(filename, fmWrite)
  defer: stream.close()
  
  let sampleData = convertTo16Bit(samples)
  let dataSize = sampleData.len * 2
  let header = createWaveHeader(sampleRate, 1, 16, dataSize)
  
  # ヘッダー書き込み
  stream.write(header.chunkID)
  stream.write(header.chunkSize)
  stream.write(header.format)
  stream.write(header.subchunk1ID)
  stream.write(header.subchunk1Size)
  stream.write(header.audioFormat)
  stream.write(header.numChannels)
  stream.write(header.sampleRate)
  stream.write(header.byteRate)
  stream.write(header.blockAlign)
  stream.write(header.bitsPerSample)
  stream.write(header.subchunk2ID)
  stream.write(header.subchunk2Size)
  
  # サンプルデータ書き込み
  for sample in sampleData:
    stream.write(sample)

# MIDIからWAV変換のメイン関数
proc convertMidiToWav(midiFilename, wavFilename: string, tempo: float): bool =
  let midiFileOpt = readMidiFile(midiFilename)
  if midiFileOpt.isNone:
    echo "MIDIファイルの読み込みに失敗しました"
    return false
  
  let midiFile = midiFileOpt.get()
  echo "MIDIファイルを読み込みました: Format ", midiFile.format, 
       ", ", midiFile.trackCount, " tracks, ", midiFile.ticksPerQuarter, " ticks/quarter"
  
  let notes = eventsToNotes(midiFile, tempo)
  echo notes.len, "個の音符を検出しました"
  
  if notes.len == 0:
    echo "音符が見つかりませんでした"
    return false
  
  let totalDuration = notes.mapIt(it.startTime + it.duration).max() + 1.0 # 余裕を持たせる
  let sampleRate = 44100
  
  echo "総演奏時間: ", totalDuration.formatFloat(ffDecimal, 2), "秒"
  echo "波形を生成中..."
  
  let waveforms = notes.mapIt(noteToWaveform(it, sampleRate, totalDuration))
  let mixed = mixWaveforms(waveforms)
  writeWaveFile(wavFilename, mixed, sampleRate)
  
  echo "WAVファイル '", wavFilename, "' を生成しました！"
  return true

# デモ用のサンプル生成
proc generateDemoNESMusic() =
  let sampleRate = 44100
  let duration = 2.0
  
  # チャンネル1: 矩形波（メロディ）
  let square1 = generateSquareWave(440.0, Duty50, sampleRate, duration)
  let square1Final = square1.mapIt(it * 0.3)
  
  # チャンネル2: 矩形波（ハーモニー）
  let square2 = generateSquareWave(330.0, Duty25, sampleRate, duration)
  let square2Final = square2.mapIt(it * 0.25)
  
  # チャンネル3: 三角波（ベース）
  let triangle = generateTriangleWave(110.0, sampleRate, duration)
  let triangleFinal = triangle.mapIt(it * 0.4)
  
  # ミックス
  let mixed = mixWaveforms(@[square1Final, square2Final, triangleFinal])
  
  # WAVファイルに出力
  writeWaveFile("demo_nes_sound.wav", mixed, sampleRate)
  echo "デモNES音源のWAVファイル 'demo_nes_sound.wav' を生成しました！"

# メイン実行部分
when isMainModule:
  try:
    let args = commandLineParams()
    if args.len >= 1:
      let midiFile = args[0]
      let outputFile = if args.len >= 2: args[1] else: changeFileExt(midiFile, "wav")
      let tempo = if args.len >= 3: parseFloat(args[2]) else: 120.0
      
      if fileExists(midiFile):
        if convertMidiToWav(midiFile, outputFile, tempo):
          quit(0)
        else:
          quit(1)
      else:
        echo "MIDIファイル '", midiFile, "' が見つかりません"
        quit(1)
    else:
      echo "使用方法: program <MIDIファイル> [出力WAVファイル] [テンポ]"
      echo "デモファイルを生成します..."
      generateDemoNESMusic()
      quit(0)
  except:
    echo "エラーが発生しました: ", getCurrentExceptionMsg()
    quit(1)

File: dart/bin/midi2chiptune.dart
// MIDIをNES(ファミコン)風チップチューンWAVに変換するCLI
// dart run midi2chiptune <MIDIファイル> [出力WAVファイル] [テンポ]

import 'dart:io';
import 'dart:math' as math;
import 'dart:typed_data';

// --- MIDI データ構造 ---

enum MidiEventType { noteOn, noteOff, programChange, unknown }

class MidiEvent {
  final MidiEventType type;
  final int channel;
  final int note;
  final int velocity;
  final int program;
  final int deltaTime;

  MidiEvent({
    required this.type,
    required this.channel,
    required this.note,
    required this.velocity,
    required this.program,
    required this.deltaTime,
  });
}

class MidiTrack {
  final List<MidiEvent> events;
  MidiTrack(this.events);
}

class MidiFile {
  final int format;
  final int trackCount;
  final int ticksPerQuarter;
  final List<MidiTrack> tracks;

  MidiFile({
    required this.format,
    required this.trackCount,
    required this.ticksPerQuarter,
    required this.tracks,
  });
}

class Note {
  final int midiNote;
  final int channel;
  final double startTime;
  final double duration;
  final int velocity;

  Note({
    required this.midiNote,
    required this.channel,
    required this.startTime,
    required this.duration,
    required this.velocity,
  });
}

// 矩形波デューティサイクル（NES 4種）
enum DutyCycle { duty12_5, duty25, duty50, duty75 }

double dutyCycleValue(DutyCycle duty) {
  switch (duty) {
    case DutyCycle.duty12_5:
      return 0.125;
    case DutyCycle.duty25:
      return 0.25;
    case DutyCycle.duty50:
      return 0.5;
    case DutyCycle.duty75:
      return 0.75;
  }
}

// バイト列リーダー（位置管理）
class ByteReader {
  final Uint8List data;
  int pos = 0;

  ByteReader(this.data);

  bool get hasMore => pos < data.length;

  int readByte() {
    if (pos >= data.length) throw StateError('EOF');
    return data[pos++];
  }

  int readVLQ() {
    int value = 0;
    int b;
    do {
      b = readByte();
      value = (value << 7) | (b & 0x7F);
    } while ((b & 0x80) != 0);
    return value;
  }

  int readUint16BE() {
    if (pos + 2 > data.length) throw StateError('EOF');
    final high = data[pos++];
    final low = data[pos++];
    return (high << 8) | low;
  }

  int readUint32BE() {
    if (pos + 4 > data.length) throw StateError('EOF');
    return (data[pos++] << 24) |
        (data[pos++] << 16) |
        (data[pos++] << 8) |
        data[pos++];
  }

  void skip(int n) {
    pos = math.min(pos + n, data.length);
  }
}

double midiNoteToFrequency(int midiNote) {
  return 440.0 * math.pow(2.0, (midiNote - 69) / 12.0).toDouble();
}

// MIDI ヘッダー読み込み
(List<int>, List<MidiTrack>) readMidiFile(Uint8List bytes) {
  final reader = ByteReader(bytes);

  if (reader.data.length < 14) throw FormatException('Invalid MIDI: too short');
  final chunkId = String.fromCharCodes(reader.data.sublist(0, 4));
  reader.pos = 4;
  if (chunkId != 'MThd') throw FormatException('Invalid MIDI header');

  reader.readUint32BE(); // chunk length
  final format = reader.readUint16BE();
  final trackCount = reader.readUint16BE();
  final ticksPerQuarter = reader.readUint16BE();

  final tracks = <MidiTrack>[];
  for (var i = 0; i < trackCount && reader.hasMore; i++) {
    if (reader.pos + 8 > reader.data.length) break;
    final track = readMidiTrack(reader);
    tracks.add(track);
  }

  return ([format, trackCount, ticksPerQuarter], tracks);
}

MidiTrack readMidiTrack(ByteReader reader) {
  final chunkType = String.fromCharCodes([
    reader.data[reader.pos],
    reader.data[reader.pos + 1],
    reader.data[reader.pos + 2],
    reader.data[reader.pos + 3],
  ]);
  reader.pos += 4;
  if (chunkType != 'MTrk') throw FormatException('Invalid track header');

  final chunkLength = reader.readUint32BE();
  final endPos = reader.pos + chunkLength;
  final events = <MidiEvent>[];
  var runningStatus = 0;

  while (reader.pos < endPos && reader.hasMore) {
    try {
      final deltaTime = reader.readVLQ();
      final (evt, newStatus) = parseMidiEvent(reader, runningStatus, deltaTime);
      runningStatus = newStatus;
      if (evt.type != MidiEventType.unknown) {
        events.add(evt);
      }
    } catch (_) {
      if (reader.pos < endPos && reader.hasMore) reader.readByte();
    }
  }
  reader.pos = endPos;

  return MidiTrack(events);
}

(MidiEvent, int) parseMidiEvent(ByteReader reader, int runningStatus, int deltaTime) {
  if (!reader.hasMore) return (_unknown(deltaTime), runningStatus);

  var status = runningStatus;
  final firstByte = reader.readByte();

  if (firstByte >= 0x80) {
    status = firstByte;
  } else {
    reader.pos--; // push back for running status data
  }

  final eventType = status & 0xF0;
  final channel = status & 0x0F;

  switch (eventType) {
    case 0x90: // Note On
      final note = reader.readByte();
      final velocity = reader.readByte();
      if (velocity == 0) {
        return (
          MidiEvent(
            type: MidiEventType.noteOff,
            channel: channel,
            note: note,
            velocity: 0,
            program: 0,
            deltaTime: deltaTime,
          ),
          status,
        );
      }
      return (
        MidiEvent(
          type: MidiEventType.noteOn,
          channel: channel,
          note: note,
          velocity: velocity,
          program: 0,
          deltaTime: deltaTime,
        ),
        status,
      );

    case 0x80: // Note Off
      final note = reader.readByte();
      reader.readByte(); // velocity
      return (
        MidiEvent(
          type: MidiEventType.noteOff,
          channel: channel,
          note: note,
          velocity: 0,
          program: 0,
          deltaTime: deltaTime,
        ),
        status,
      );

    case 0xC0: // Program Change
      final program = reader.readByte();
      return (
        MidiEvent(
          type: MidiEventType.programChange,
          channel: channel,
          note: 0,
          velocity: 0,
          program: program,
          deltaTime: deltaTime,
        ),
        status,
      );

    default:
      if (status == 0xFF) {
        reader.readByte(); // meta type
        final length = reader.readVLQ();
        reader.skip(length);
      } else if (status >= 0x80) {
        reader.readByte();
        if ((status & 0xE0) != 0xC0 && (status & 0xE0) != 0xD0) {
          if (reader.hasMore) reader.readByte();
        }
      }
      return (_unknown(deltaTime), status);
  }
}

MidiEvent _unknown(int deltaTime) => MidiEvent(
      type: MidiEventType.unknown,
      channel: 0,
      note: 0,
      velocity: 0,
      program: 0,
      deltaTime: deltaTime,
    );

// イベント → 音符リスト
List<Note> eventsToNotes(MidiFile midiFile, double tempo) {
  double ticksToSeconds(int tick) =>
      tick / midiFile.ticksPerQuarter * 60.0 / tempo;

  final noteOnMap = <String, ({int velocity, double startTime})>{};
  final notes = <Note>[];

  for (final track in midiFile.tracks) {
    var currentTick = 0;
    for (final evt in track.events) {
      currentTick += evt.deltaTime;
      final key = '${evt.channel}-${evt.note}';

      switch (evt.type) {
        case MidiEventType.noteOn:
          noteOnMap[key] = (
            velocity: evt.velocity,
            startTime: ticksToSeconds(currentTick),
          );
          break;
        case MidiEventType.noteOff:
          final onInfo = noteOnMap.remove(key);
          if (onInfo != null) {
            final endTime = ticksToSeconds(currentTick);
            final duration = endTime - onInfo.startTime;
            if (duration > 0) {
              notes.add(Note(
                midiNote: evt.note,
                channel: evt.channel,
                startTime: onInfo.startTime,
                duration: duration,
                velocity: onInfo.velocity,
              ));
            }
          }
          break;
        default:
          break;
      }
    }
  }
  return notes;
}

// --- 波形生成 ---

List<double> generateSquareWave(
    double frequency, DutyCycle duty, int sampleRate, double duration) {
  final samples = (sampleRate * duration).floor();
  final dutyVal = dutyCycleValue(duty);
  return List.generate(samples, (i) {
    final t = i / sampleRate;
    final phase = (t * frequency) % 1.0;
    return phase < dutyVal ? 1.0 : -1.0;
  });
}

List<double> generateTriangleWave(
    double frequency, int sampleRate, double duration) {
  final samples = (sampleRate * duration).floor();
  return List.generate(samples, (i) {
    final t = i / sampleRate;
    final phase = (t * frequency) % 1.0;
    return phase < 0.5 ? 4.0 * phase - 1.0 : 3.0 - 4.0 * phase;
  });
}

List<double> generateNoise(bool isShort, int sampleRate, double duration) {
  final samples = (sampleRate * duration).floor();
  var shift = 1;
  return List.generate(samples, (_) {
    final bit0 = shift & 1;
    final bit1 = isShort ? (shift >> 6) & 1 : (shift >> 1) & 1;
    final feedback = bit0 ^ bit1;
    shift = (shift >> 1) | (feedback << 14);
    return bit0 == 1 ? 1.0 : -1.0;
  });
}

List<double> noteToWaveform(
    Note note, int sampleRate, double totalDuration) {
  final frequency = midiNoteToFrequency(note.midiNote);
  final volume = (note.velocity / 127.0) * 0.7;
  final startSample = (note.startTime * sampleRate).floor();
  final noteSamples = (note.duration * sampleRate).floor();
  final totalSamples = (totalDuration * sampleRate).floor();

  List<double> waveform;
  switch (note.channel % 4) {
    case 0:
      waveform =
          generateSquareWave(frequency, DutyCycle.duty50, sampleRate, note.duration);
      break;
    case 1:
      waveform =
          generateSquareWave(frequency, DutyCycle.duty25, sampleRate, note.duration);
      break;
    case 2:
      waveform = generateTriangleWave(frequency, sampleRate, note.duration);
      break;
    default:
      waveform = generateNoise(false, sampleRate, note.duration);
  }

  final result = List.filled(totalSamples, 0.0);
  final endSample = math.min(startSample + noteSamples, totalSamples);
  for (var i = startSample; i < endSample && i - startSample < waveform.length; i++) {
    result[i] = waveform[i - startSample] * volume;
  }
  return result;
}

List<double> mixWaveforms(List<List<double>> waveforms) {
  if (waveforms.isEmpty) return [];
  final maxLength = waveforms.map((w) => w.length).reduce(math.max);
  final n = waveforms.length.toDouble();
  return List.generate(maxLength, (i) {
    var sum = 0.0;
    for (final w in waveforms) {
      if (i < w.length) sum += w[i];
    }
    return sum / n;
  });
}

Uint8List createWaveHeader(int sampleRate, int numChannels, int bitsPerSample, int dataSize) {
  final byteRate = sampleRate * numChannels * (bitsPerSample ~/ 8);
  final blockAlign = numChannels * (bitsPerSample ~/ 8);
  final header = ByteData(44);
  var o = 0;
  header.setUint8(o++, 0x52); header.setUint8(o++, 0x49);
  header.setUint8(o++, 0x46); header.setUint8(o++, 0x46);
  header.setUint32(o, 36 + dataSize, Endian.little); o += 4;
  header.setUint8(o++, 0x57); header.setUint8(o++, 0x41);
  header.setUint8(o++, 0x56); header.setUint8(o++, 0x45);
  header.setUint8(o++, 0x66); header.setUint8(o++, 0x6d);
  header.setUint8(o++, 0x74); header.setUint8(o++, 0x20);
  header.setUint32(o, 16, Endian.little); o += 4;
  header.setUint16(o, 1, Endian.little); o += 2;
  header.setUint16(o, numChannels, Endian.little); o += 2;
  header.setUint32(o, sampleRate, Endian.little); o += 4;
  header.setUint32(o, byteRate, Endian.little); o += 4;
  header.setUint16(o, blockAlign, Endian.little); o += 2;
  header.setUint16(o, bitsPerSample, Endian.little); o += 2;
  header.setUint8(o++, 0x64); header.setUint8(o++, 0x61);
  header.setUint8(o++, 0x74); header.setUint8(o++, 0x61);
  header.setUint32(o, dataSize, Endian.little);
  return header.buffer.asUint8List();
}

void writeWaveFile(String filename, List<double> samples, int sampleRate) {
  final sampleData = samples
      .map((s) {
        final clamped = s.clamp(-1.0, 1.0);
        return (clamped * 32767).round().clamp(-32768, 32767);
      })
      .toList();
  final dataSize = sampleData.length * 2;
  final header = createWaveHeader(sampleRate, 1, 16, dataSize);
  final file = File(filename).openSync(mode: FileMode.write);
  file.writeFromSync(header);
  final data = ByteData(sampleData.length * 2);
  for (var i = 0; i < sampleData.length; i++) {
    data.setInt16(i * 2, sampleData[i], Endian.little);
  }
  file.writeFromSync(data.buffer.asUint8List());
  file.closeSync();
}

bool convertMidiToWav(String midiPath, String wavPath, double tempo) {
  try {
    final bytes = File(midiPath).readAsBytesSync();
    final (header, tracks) = readMidiFile(bytes);
    final format = header[0];
    final trackCount = header[1];
    final ticksPerQuarter = header[2];
    final midiFile = MidiFile(
      format: format,
      trackCount: trackCount,
      ticksPerQuarter: ticksPerQuarter,
      tracks: tracks,
    );

    print('MIDIファイルを読み込みました: Format $format, $trackCount tracks, $ticksPerQuarter ticks/quarter');

    final notes = eventsToNotes(midiFile, tempo);
    print('${notes.length}個の音符を検出しました');

    if (notes.isEmpty) {
      print('音符が見つかりませんでした');
      return false;
    }

    double totalDuration = 0;
    for (final n in notes) {
      final end = n.startTime + n.duration;
      if (end > totalDuration) totalDuration = end;
    }
    totalDuration += 1.0;

    const sampleRate = 44100;
    print('総演奏時間: ${totalDuration.toStringAsFixed(2)}秒');
    print('波形を生成中...');

    final waveforms = notes.map((n) => noteToWaveform(n, sampleRate, totalDuration)).toList();
    final mixed = mixWaveforms(waveforms);
    writeWaveFile(wavPath, mixed, sampleRate);
    print("WAVファイル '$wavPath' を生成しました！");
    return true;
  } catch (e, st) {
    print('変換エラー: $e');
    print(st);
    return false;
  }
}

void generateDemoNESMusic() {
  const sampleRate = 44100;
  const duration = 2.0;

  final square1 = generateSquareWave(440.0, DutyCycle.duty50, sampleRate, duration)
      .map((s) => s * 0.3)
      .toList();
  final square2 = generateSquareWave(330.0, DutyCycle.duty25, sampleRate, duration)
      .map((s) => s * 0.25)
      .toList();
  final triangle = generateTriangleWave(110.0, sampleRate, duration)
      .map((s) => s * 0.4)
      .toList();

  final mixed = mixWaveforms([square1, square2, triangle]);
  writeWaveFile('demo_nes_sound.wav', mixed, sampleRate);
  print("デモNES音源のWAVファイル 'demo_nes_sound.wav' を生成しました！");
}

void main(List<String> args) {
  if (args.isNotEmpty) {
    final midiFile = args[0];
    final outputFile = args.length >= 2
        ? args[1]
        : midiFile.replaceAll(RegExp(r'\.mid$'), '.wav');
    final tempo = args.length >= 3 ? double.tryParse(args[2]) ?? 120.0 : 120.0;

    if (!File(midiFile).existsSync()) {
      print("MIDIファイル '$midiFile' が見つかりません");
      exit(1);
    }
    exit(convertMidiToWav(midiFile, outputFile, tempo) ? 0 : 1);
  } else {
    print('使用方法: dart run midi2chiptune <MIDIファイル> [出力WAVファイル] [テンポ]');
    print('デモファイルを生成します...');
    generateDemoNESMusic();
  }
}
